File: apps/api/src/modules/auth/auth.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Post,
  Req,
  Res,
} from '@nestjs/common';
import {
  ApiConflictResponse,
  ApiCreatedResponse,
  ApiOkResponse,
  ApiOperation,
  ApiTags,
  ApiTooManyRequestsResponse,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { Throttle } from '@nestjs/throttler';
import type { AuthSessionResponse, AuthenticatedUser } from '@gamescore/types';
import type { Request, Response } from 'express';

import { AppConfigService } from '../../common/config/app-config.service';
import { UnauthorizedError } from '../../common/errors/app.exception';
import { ERROR_CODES } from '../../common/errors/error-codes';
import { AuthService, type AuthResult } from './application/auth.service';
import { Authenticated } from './decorators/auth.decorators';
import { CurrentUser } from './decorators/current-user.decorator';
import { REFRESH_TOKEN_COOKIE } from './domain/auth-user';
import type { AuthUser } from './domain/auth-user';
import { LoginDto } from './dto/login.dto';
import { RegisterDto } from './dto/register.dto';
import { ChangePasswordDto, ForgotPasswordDto, ResetPasswordDto, ResendVerificationDto, VerifyEmailDto } from './dto/password.dto';

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(
    private readonly auth: AuthService,
    private readonly config: AppConfigService,
  ) {}

  @Post('register')
  @HttpCode(HttpStatus.CREATED)
  // Registration is the most abused endpoint of any public site, so it is the
  // most tightly limited one.
  @Throttle({ default: { limit: 5, ttl: 60 * 60 * 1000 } })
  @ApiOperation({ summary: 'Create an account and start a session' })
  @ApiCreatedResponse({ description: 'Account created and signed in' })
  @ApiConflictResponse({ description: 'Email or username already in use' })
  @ApiTooManyRequestsResponse({ description: 'Too many registrations from this client' })
  async register(
    @Body() dto: RegisterDto,
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthSessionResponse> {
    const result = await this.auth.register(dto, request.headers['user-agent']);
    return this.completeSession(response, result);
  }

  @Post('login')
  @HttpCode(HttpStatus.OK)
  @Throttle({ default: { limit: 10, ttl: 60 * 1000, blockDuration: 5 * 60 * 1000 } })
  @ApiOperation({ summary: 'Sign in with email or username' })
  @ApiOkResponse({ description: 'Signed in' })
  @ApiUnauthorizedResponse({ description: 'Invalid credentials' })
  @ApiTooManyRequestsResponse({ description: 'Too many attempts; try again later' })
  async login(
    @Body() dto: LoginDto,
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthSessionResponse> {
    const result = await this.auth.login(dto, request.headers['user-agent']);
    return this.completeSession(response, result);
  }

  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  @Throttle({ default: { limit: 60, ttl: 60 * 60 * 1000 } })
  @ApiOperation({
    summary: 'Exchange the refresh cookie for a new access token',
    description:
      'Reads the httpOnly refresh cookie, rotates it, and returns a fresh access token. ' +
      'Replaying an already rotated token revokes every session for that account.',
  })
  @ApiOkResponse({ description: 'New access token issued' })
  @ApiUnauthorizedResponse({ description: 'Missing, expired or already used refresh token' })
  async refresh(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthSessionResponse> {
    const token = this.readRefreshCookie(request);
    if (!token) {
      throw new UnauthorizedError(ERROR_CODES.INVALID_REFRESH_TOKEN, 'No refresh token provided');
    }

    try {
      const result = await this.auth.refresh(token, request.headers['user-agent']);
      return this.completeSession(response, result);
    } catch (error) {
      // The cookie is useless now, so do not leave it in the browser.
      this.clearRefreshCookie(response);
      throw error;
    }
  }

  @Post('logout')
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Revoke the current refresh token and clear the cookie' })
  async logout(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<void> {
    await this.auth.logout(this.readRefreshCookie(request));
    this.clearRefreshCookie(response);
  }

  @Get('me')
  @Authenticated()
  @ApiOperation({ summary: 'The signed-in account' })
  @ApiOkResponse({ description: 'Current account' })
  async me(@CurrentUser() user: AuthUser): Promise<AuthenticatedUser> {
    return this.auth.currentUser(user.id);
  }

  @Post('password')
  @Authenticated()
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiOperation({ summary: 'Change the signed-in password' })
  async changePassword(
    @CurrentUser() user: AuthUser,
    @Body() dto: ChangePasswordDto,
  ): Promise<void> {
    await this.auth.changePassword(user.id, dto.currentPassword, dto.newPassword);
  }

  @Post('forgot-password')
  @HttpCode(HttpStatus.NO_CONTENT)
  @Throttle({ default: { limit: 5, ttl: 60 * 1000, blockDuration: 5 * 60 * 1000 } })
  @ApiOperation({ summary: 'Request a password reset email. Always succeeds.' })
  async forgotPassword(@Body() dto: ForgotPasswordDto): Promise<void> {
    await this.auth.forgotPassword(dto.email);
  }

  @Post('reset-password')
  @HttpCode(HttpStatus.NO_CONTENT)
  @Throttle({ default: { limit: 10, ttl: 60 * 1000, blockDuration: 5 * 60 * 1000 } })
  @ApiOperation({ summary: 'Set a new password using a reset token' })
  async resetPassword(@Body() dto: ResetPasswordDto): Promise<void> {
    await this.auth.resetPassword(dto.token, dto.password);
  }

  @Post('verify-email')
  @HttpCode(HttpStatus.NO_CONTENT)
  @Throttle({ default: { limit: 10, ttl: 60 * 1000, blockDuration: 5 * 60 * 1000 } })
  @ApiOperation({ summary: 'Confirm the account email with a verification token' })
  async verifyEmail(@Body() dto: VerifyEmailDto): Promise<void> {
    await this.auth.verifyEmail(dto.token);
  }

  @Post('resend-verification')
  @HttpCode(HttpStatus.NO_CONTENT)
  @Throttle({ default: { limit: 5, ttl: 60 * 1000, blockDuration: 5 * 60 * 1000 } })
  @ApiOperation({ summary: 'Resend the verification email. Always succeeds.' })
  async resendVerification(@Body() dto: ResendVerificationDto): Promise<void> {
    await this.auth.resendVerification(dto.email);
  }

  private completeSession(response: Response, result: AuthResult): AuthSessionResponse {
    this.setRefreshCookie(response, result.refreshToken.token, result.refreshToken.expiresAt);
    return result.session;
  }

  /**
   * The refresh token lives in an httpOnly cookie so no script can read it,
   * while the access token stays in memory in the browser. `sameSite: lax` is
   * enough here because the web app and the API share the `localhost` site
   * (cookie scope ignores the port).
   */
  private setRefreshCookie(response: Response, token: string, expiresAt: Date): void {
    response.cookie(REFRESH_TOKEN_COOKIE, token, {
      httpOnly: true,
      sameSite: 'lax',
      secure: this.config.isProduction,
      path: '/',
      expires: expiresAt,
    });
  }

  private clearRefreshCookie(response: Response): void {
    response.clearCookie(REFRESH_TOKEN_COOKIE, {
      httpOnly: true,
      sameSite: 'lax',
      secure: this.config.isProduction,
      path: '/',
    });
  }

  private readRefreshCookie(request: Request): string | undefined {
    const cookies = request.cookies as Record<string, string> | undefined;
    return cookies?.[REFRESH_TOKEN_COOKIE];
  }
}

File: apps/api/src/GameScore.Api/Controllers/AuthController.cs
using GameScore.Api.Authorization;
using GameScore.Application.Auth;
using GameScore.Application.Configuration;
using GameScore.Infrastructure.Auth;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace GameScore.Api.Controllers;

[ApiController]
[Route("auth")]
public sealed class AuthController(AuthService auth, IOptions<AppConfig> config) : ControllerBase
{
    [HttpPost("register")]
    [ProducesResponseType(typeof(AuthSessionResponse), StatusCodes.Status201Created)]
    public async Task<ActionResult<AuthSessionResponse>> Register([FromBody] RegisterRequest body, CancellationToken cancellationToken)
    {
        var result = await auth.RegisterAsync(
            body.Email,
            body.Username,
            body.Password,
            body.DisplayName,
            Request.Headers.UserAgent,
            cancellationToken);
        SetRefreshCookie(result.RefreshToken);
        return Created(string.Empty, result.Session);
    }

    [HttpPost("login")]
    public async Task<ActionResult<AuthSessionResponse>> Login([FromBody] LoginRequest body, CancellationToken cancellationToken)
    {
        var result = await auth.LoginAsync(body.Identifier, body.Password, Request.Headers.UserAgent, cancellationToken);
        SetRefreshCookie(result.RefreshToken);
        return Ok(result.Session);
    }

    [HttpPost("refresh")]
    public async Task<ActionResult<AuthSessionResponse>> Refresh(CancellationToken cancellationToken)
    {
        if (!Request.Cookies.TryGetValue(AuthConstants.RefreshTokenCookie, out var token))
        {
            throw new Application.Errors.UnauthorizedException(
                Application.Errors.ErrorCodes.InvalidRefreshToken,
                "No refresh token provided");
        }

        try
        {
            var result = await auth.RefreshAsync(token, Request.Headers.UserAgent, cancellationToken);
            SetRefreshCookie(result.RefreshToken);
            return Ok(result.Session);
        }
        catch (Application.Errors.UnauthorizedException)
        {
            ClearRefreshCookie();
            throw;
        }
    }

    [HttpPost("logout")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> Logout(CancellationToken cancellationToken)
    {
        Request.Cookies.TryGetValue(AuthConstants.RefreshTokenCookie, out var token);
        await auth.LogoutAsync(token, cancellationToken);
        ClearRefreshCookie();
        return NoContent();
    }

    [HttpGet("me")]
    [Microsoft.AspNetCore.Authorization.Authorize]
    public async Task<ActionResult<AuthenticatedUserDto>> Me(CancellationToken cancellationToken)
    {
        var user = HttpContext.User.GetAuthUser()
            ?? throw new Application.Errors.UnauthorizedException();
        return Ok(await auth.CurrentUserAsync(user.Id, cancellationToken));
    }

    [HttpPost("password")]
    [Microsoft.AspNetCore.Authorization.Authorize]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordRequest body, CancellationToken cancellationToken)
    {
        var user = HttpContext.User.GetAuthUser()
            ?? throw new Application.Errors.UnauthorizedException();
        await auth.ChangePasswordAsync(user.Id, body.CurrentPassword, body.NewPassword, cancellationToken);
        return NoContent();
    }

    [HttpPost("forgot-password")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordRequest body, CancellationToken cancellationToken)
    {
        await auth.ForgotPasswordAsync(body.Email, cancellationToken);
        return NoContent();
    }

    [HttpPost("reset-password")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordRequest body, CancellationToken cancellationToken)
    {
        await auth.ResetPasswordAsync(body.Token, body.Password, cancellationToken);
        return NoContent();
    }

    [HttpPost("verify-email")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> VerifyEmail([FromBody] VerifyEmailRequest body, CancellationToken cancellationToken)
    {
        await auth.VerifyEmailAsync(body.Token, cancellationToken);
        return NoContent();
    }

    [HttpPost("resend-verification")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    public async Task<IActionResult> ResendVerification([FromBody] ResendVerificationRequest body, CancellationToken cancellationToken)
    {
        await auth.ResendVerificationAsync(body.Email, cancellationToken);
        return NoContent();
    }

    private void SetRefreshCookie(IssuedRefreshToken refresh)
    {
        Response.Cookies.Append(
            AuthConstants.RefreshTokenCookie,
            refresh.Token,
            new CookieOptions
            {
                HttpOnly = true,
                SameSite = SameSiteMode.Lax,
                Secure = config.Value.IsProduction,
                Path = "/",
                Expires = refresh.ExpiresAt,
            });
    }

    private void ClearRefreshCookie()
    {
        Response.Cookies.Delete(
            AuthConstants.RefreshTokenCookie,
            new CookieOptions
            {
                HttpOnly = true,
                SameSite = SameSiteMode.Lax,
                Secure = config.Value.IsProduction,
                Path = "/",
            });
    }
}

public sealed record RegisterRequest(string Email, string Username, string Password, string? DisplayName, bool AcceptedTerms);
public sealed record LoginRequest(string Identifier, string Password);
public sealed record ChangePasswordRequest(string CurrentPassword, string NewPassword);
public sealed record ForgotPasswordRequest(string Email);
public sealed record ResetPasswordRequest(string Token, string Password);
public sealed record VerifyEmailRequest(string Token);
public sealed record ResendVerificationRequest(string Email);
